### ISSUE TITLE
False positive when assigning an empty dict to a union of `dict`s or `Mapping`s

### ISSUE THREAD

**Issue by hamdanal** [python/mypy#16044]
**Bug Report**

Like the title says.
I thought it might be because of dict variance but mapping union errors as well.

**To Reproduce**

```python
from collections.abc import Mapping

d1: dict[str, str] | dict[int, int] = {}
d2: Mapping[str, str] | Mapping[int, int] = {}

reveal_type(d1)
reveal_type(d2)
```

**Expected Behavior**

No errors.

**Actual Behavior**

```console
$ mypy t.py
t.py:3: error: Incompatible types in assignment (expression has type "dict[<nothing>, <nothing>]", variable has type "dict[str, str] | dict[int, int]")  [assignment]
t.py:4: error: Incompatible types in assignment (expression has type "dict[<nothing>, <nothing>]", variable has type "Mapping[str, str] | Mapping[int, int]")  [assignment]
t.py:6: note: Revealed type is "Union[builtins.dict[builtins.str, builtins.str], builtins.dict[builtins.int, builtins.int]]"
t.py:7: note: Revealed type is "Union[typing.Mapping[builtins.str, builtins.str], typing.Mapping[builtins.int, builtins.int]]"
```

**Your Environment**

- Mypy version used: mypy 1.5.1 (compiled: yes)
- Mypy command-line flags: N/A
- Mypy configuration options from `mypy.ini` (and other config files): N/A
- Python version used: 3.10

**Comment by ilevkivskyi** [comment]
A duplicate of https://github.com/python/mypy/issues/6463
